Add Web Extension support for loading images and parsing `icons`, `action`, `browser_action`, and `page_action`.

xeenon opened the pull request:
#### 9aac1c9848c4cecd92016e689a4bb7af115b2abd
<pre>
Add Web Extension support for loading images and parsing `icons`, `action`, `browser_action`, and `page_action`.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246168">https://bugs.webkit.org/show_bug.cgi?id=246168</a>

Reviewed by Brian Weinstein.

Added support for loading and validating resources from the extension. This is now used for loading the
manifest.json file and also icons. The icons are generated using the platform&apos;s image class. On macOS
the 1x and 2x sizes are retrieved and put into a single NSImage. On iOS the largest scale factor is used
based on the main UIScreen scale. SVG images are also supported via some SPIs.

This loading mechanism is now also used for the special &quot;_generated_background_page.html&quot; path. It will also
be used to load content scripts and stylesheets for injected content.

* Source/WTF/wtf/PlatformUse.h: Added USE_NSIMAGE_FOR_SVG_SUPPORT define.
* Source/WTF/wtf/spi/cocoa/SecuritySPI.h: Added SecCodeValidateFileResource.
* Source/WebCore/PAL/pal/spi/mac/NSImageSPI.h: Added _NSSVGImageRep class.
* Source/WebCore/en.lproj/Localizable.strings: Updated with update-webkit-localizable-strings.
* Source/WebCore/platform/LocalizedStrings.h: Exported formatLocalizedString to use in WebKit.
* Source/WebKit/Platform/spi/Cocoa/CoreSVGSPI.h: Added.
* Source/WebKit/Platform/spi/Cocoa/FoundationSPI.h: Added.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h: Added _imageWithCGSVGDocument: method to UIImage.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
Simplify the exposed enum values for _WKWebExtensionError to reduce API surface and stay nimble.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension _initWithManifestDictionary:]):
(-[_WKWebExtension _initWithManifestDictionary:resources:]): Added.
(-[_WKWebExtension _initWithResources:]): Renamed from _initWithManifestData:.
(-[_WKWebExtension displayActionLabel]): Added.
(-[_WKWebExtension iconForSize:]): Added.
(-[_WKWebExtension actionIconForSize:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::bundleStaticCode): Added.
(WebKit::WebExtension::validateResourceData): Added.
(WebKit::WebExtension::resourceFileURLForPath): Added.
(WebKit::WebExtension::resourceDataForPath): Added.
(WebKit::WebExtension::createError): Tweaked strings and API enum values used.
(WebKit::WebExtension::errors): Call populateActionPropertiesIfNeeded.
(WebKit::WebExtension::icon): Added.
(WebKit::WebExtension::actionIcon): Added.
(WebKit::WebExtension::displayActionLabel): Added.
(WebKit::WebExtension::actionPopupPath): Added.
(WebKit::WebExtension::populateActionPropertiesIfNeeded): Added.
(WebKit::WebExtension::imageForPath): Added.
(WebKit::WebExtension::pathForBestImageInIconsDictionary): Added.
(WebKit::WebExtension::bestImageInIconsDictionary): Added.
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey): Added.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
 Added ActionParsing test and tweaks error enum names.

Canonical link: <a href="https://commits.webkit.org/255284@main">https://commits.webkit.org/255284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6f13df285eab43ee6dcfae16ceef26f40023960

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1266 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101830 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96024 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1260 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97682 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78566 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/83368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/36103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78435 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33846 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/27091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37721 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81057 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1653 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39607 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/17828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
```diff
@@ -36,21 +36,13 @@ WK_EXTERN NSErrorDomain const _WKWebExtensionErrorDomain WK_API_AVAILABLE(macos(
-    _WKWebExtensionErrorInvalidName,
-    _WKWebExtensionErrorInvalidURLOverrides,
-    _WKWebExtensionErrorInvalidVersion,
-    _WKWebExtensionErrorInvalidWebAccessibleResources,
```
It looks like we still kept around some localizable strings. Were these all made one error domain/code?

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
```diff
@@ -68,25 +68,30 @@ - (instancetype)initWithResourceBaseURL:(NSURL *)resourceBaseURL
         return nil;
 
-    API::Object::constructInWrapper<WebKit::WebExtension>(self, manifestData);
+    API::Object::constructInWrapper<WebKit::WebExtension>(self, resources);
```
Do we need to pass a nil manifest here? Is this constructor smart enough to figure that out?

xeenon replied:
It now looks for the `manifest.json` entry in the resources dictionary.

xeenon replied:
The `WebExtension::WebExtension` constructor also changed.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -130,13 +153,9 @@
-        recordError(createError(Error::ManifestNotFound));
+    NSData *manifestData = resourceDataForPath(@"manifest.json");
+    if (!manifestData)
         return nil;
```
Do we want any of the errors here? Or are these handled elsewhere?

xeenon replied:
It is handled by the more generic `Error::ResourceNotFound` error and the `resourceDataForPath()` method.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -29,6 +29,7 @@
 
 #include "APIObject.h"
 #include "WebExtensionMatchPattern.h"
+#include <wtf/CompletionHandler.h>
```
This isn't needed. I'll remove it.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -156,6 +175,117 @@
+    return result == noErr;
+}
+
+#endif
```
Should we put a comment describing which condition this is ending?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -193,23 +323,34 @@
     case Error::InvalidBackgroundContent:
-        errorCode = _WKWebExtensionErrorInvalidBackgroundContent;
-        localizedDescription = WEB_UI_STRING("Empty or invalid `background` manifest entry.", "WKWebExtensionErrorInvalidBackgroundContent description");
+        errorCode = _WKWebExtensionErrorInvalidManifestEntry;
```
Ah I see - you replaced all of the individual error codes with ...InvalidManifestEntry.

xeenon replied:
Yep

pxlcoder commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -59,19 +70,29 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension> {
+#if PLATFORM(IOS_FAMILY)
+    using PlatformImage = UIImage;
+#endif
+
```
We can use the existing `CocoaImage` instead of adding this.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
```diff
@@ -134,11 +139,26 @@ - (NSString *)displayDescription
     return _webExtension->version();
 }
 
+- (nullable WebKit::WebExtension::PlatformImage *)iconForSize:(CGSize)size
```
We don't normally include nullability annotations in implementation files.

(5 other instances below)

xeenon replied:
Oops, yep!

xeenon committed:
Add Web Extension support for loading images and parsing `icons`, `action`, `browser_action`, and `page_action`.
https://bugs.webkit.org/show_bug.cgi?id=246168

Reviewed by Brian Weinstein.

Added support for loading and validating resources from the extension. This is now used for loading the
manifest.json file and also icons. The icons are generated using the platform's image class. On macOS
the 1x and 2x sizes are retrieved and put into a single NSImage. On iOS the largest scale factor is used
based on the main UIScreen scale. SVG images are also supported via some SPIs.

This loading mechanism is now also used for the special "_generated_background_page.html" path. It will also
be used to load content scripts and stylesheets for injected content.

* Source/WTF/wtf/PlatformUse.h: Added USE_NSIMAGE_FOR_SVG_SUPPORT define.
* Source/WTF/wtf/spi/cocoa/SecuritySPI.h: Added SecCodeValidateFileResource.
* Source/WebCore/PAL/pal/spi/mac/NSImageSPI.h: Added _NSSVGImageRep class.
* Source/WebCore/en.lproj/Localizable.strings: Updated with update-webkit-localizable-strings.
* Source/WebCore/platform/LocalizedStrings.h: Exported formatLocalizedString to use in WebKit.
* Source/WebKit/Platform/spi/Cocoa/CoreSVGSPI.h: Added.
* Source/WebKit/Platform/spi/Cocoa/FoundationSPI.h: Added.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h: Added _imageWithCGSVGDocument: method to UIImage.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
Simplify the exposed enum values for _WKWebExtensionError to reduce API surface and stay nimble.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension _initWithManifestDictionary:]):
(-[_WKWebExtension _initWithManifestDictionary:resources:]): Added.
(-[_WKWebExtension _initWithResources:]): Renamed from _initWithManifestData:.
(-[_WKWebExtension displayActionLabel]): Added.
(-[_WKWebExtension iconForSize:]): Added.
(-[_WKWebExtension actionIconForSize:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::bundleStaticCode): Added.
(WebKit::WebExtension::validateResourceData): Added.
(WebKit::WebExtension::resourceFileURLForPath): Added.
(WebKit::WebExtension::resourceDataForPath): Added.
(WebKit::WebExtension::createError): Tweaked strings and API enum values used.
(WebKit::WebExtension::errors): Call populateActionPropertiesIfNeeded.
(WebKit::WebExtension::icon): Added.
(WebKit::WebExtension::actionIcon): Added.
(WebKit::WebExtension::displayActionLabel): Added.
(WebKit::WebExtension::actionPopupPath): Added.
(WebKit::WebExtension::populateActionPropertiesIfNeeded): Added.
(WebKit::WebExtension::imageForPath): Added.
(WebKit::WebExtension::pathForBestImageInIconsDictionary): Added.
(WebKit::WebExtension::bestImageInIconsDictionary): Added.
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey): Added.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
 Added ActionParsing test and tweaks error enum names.

Canonical link: https://commits.webkit.org/255284@main
